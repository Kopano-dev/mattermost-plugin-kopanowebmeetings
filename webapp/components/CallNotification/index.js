export {default, CallNotification} from './CallNotification.jsx';