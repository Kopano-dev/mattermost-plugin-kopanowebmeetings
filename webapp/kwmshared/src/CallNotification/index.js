export {default, CallNotification} from './CallNotification';