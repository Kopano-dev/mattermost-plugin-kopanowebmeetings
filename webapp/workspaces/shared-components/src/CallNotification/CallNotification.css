.kwm-call-notification {
	position: fixed;
	width: 250px;
	padding: 0;
	color: white;
	background: rgba(56, 75, 96, 0.8);
	margin: auto;
	left: 0;
	right: 0;
	top: 50%;
	margin-top: -175px;
	z-index: 1000;
}
.kwm-call-notification .kwm-call-notification-header {
	text-transform: uppercase;
	padding: 18px;
	border-bottom: 1px solid white;
}
.kwm-call-notification .kwm-call-notification-header svg {
	fill: white;
	height: 12px;
	padding-right: 6px;
	position: relative;
	top: 1px
}
.kwm-call-notification .kwm-profile-image-container {
	margin: 18px auto;
	position: relative;
	border-radius: 50%;
	overflow: hidden;
	width: 50px;
	height: 50px;
	text-align: center;
}
.kwm-call-notification .kwm-profile-image-container img {
	height: 50px;
}
.kwm-call-notification .kwm-profile-image-container .k-cn-avatar {
	height: 50px;
	width: 50px;
	background: #c62911;
	color: white;
	font-size: 25px;
	font-weight: bold;
	line-height: 50px;
	border-radius: 50%;
}
.kwm-call-notification .kwm-call-notification-msg {
	text-transform: uppercase;
	margin-bottom: 18px;
	font-size: 1.3em;
	text-align: center;
}
.kwm-call-notification .kwm-call-buttons {
	clear: both;
	text-align: center;
	padding: 18px;
	font-size: 0;
	line-height: 0;
}
.kwm-call-notification .kwm-call-buttons button {
	margin-right: 18px;
}
.kwm-call-notification .kwm-call-buttons button:last-child {
	margin-right: 0;
}
