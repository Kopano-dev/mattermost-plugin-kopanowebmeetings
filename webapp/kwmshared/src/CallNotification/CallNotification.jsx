import React from 'react';
import PropTypes from 'prop-types';

import './CallNotification.css';
import ring from './sounds/ring.mp3';

import {PickupCallButton, HangupCallButton} from '../buttons';
import {cameraOutlined} from '../icons';

/**
 * Creates an element that can be used as image.
 *
 * @param {string} imgUrl Url of an image that can be used as avatar
 * @param {string} userName Full name of the user
 * @return {ReactElement} Can be an img element if an imgUrl is given, or a div
 * with initials of the user otherwise.
 */
const createImageElement = (imgUrl, userName) => {
	if ( imgUrl ) {
		return <img src={imgUrl} />;
	}

	// TODO: We should use an avatar component for this
	let initials = userName.split(/\s+/);
	if ( initials.length > 2 ) {
		initials.length = 2;
	}
	initials = initials.map(name => name[0].toUpperCase()).join('');

	return <div className='k-cn-avatar'>{initials}</div>;
};

const propTypes = {
	open: PropTypes.bool.isRequired,
	title: PropTypes.string,
	userName: PropTypes.string.isRequired,
	imgUrl: PropTypes.string,
	onAccept: PropTypes.func.isRequired,
	onReject: PropTypes.func.isRequired,
};
const defaultProps = {
	title: 'Incoming call',
};

const CallNotification = props => {
	if ( !props.open ) {
		return null;
	}

	const img = createImageElement(props.imgUrl, props.userName);

	return (
		<div className='kwm-call-notification'>
			<div className='kwm-call-notification-header'>
				{cameraOutlined()}
				{props.title}
			</div>
			<div className='kwm-call-notification-body'>
				<div className='kwm-profile-image-container'>
					{img}
				</div>
				<div className='kwm-call-notification-msg'>
					{props.userName}
				</div>
			</div>
			<div className='kwm-call-buttons'>
				<PickupCallButton onClick={props.onAccept} />
				<HangupCallButton onClick={props.onReject} />
			</div>
			<audio src={ring} autoPlay={true} loop={true} />
		</div>
	);
};
CallNotification.propTypes = propTypes;
CallNotification.defaultProps = defaultProps;

export default CallNotification;