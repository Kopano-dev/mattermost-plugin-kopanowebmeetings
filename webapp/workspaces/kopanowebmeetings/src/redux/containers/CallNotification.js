import {connect} from 'react-redux';

import CallNotification from '@kopanowebmeetings/shared-components/src/CallNotification';
import {selectors} from 'redux/ducks';
import * as Utils from 'utils/utils.js';

const {isOpen, getCaller} = selectors;

const mapStateToProps = (state, ownProps) => {
	const calledBy = getCaller(state['plugins-kopanowebmeetings']);
	return {
		open: isOpen(state['plugins-kopanowebmeetings']),
		userName: calledBy ? Utils.getDisplayName(calledBy) : '',
		imgUrl: calledBy ? Utils.imageURLForUser(calledBy) : '',
		onAccept: () => ownProps.onAccept(calledBy),
		onReject: () => ownProps.onReject(calledBy),
	};
};

export default connect(mapStateToProps)(CallNotification);
