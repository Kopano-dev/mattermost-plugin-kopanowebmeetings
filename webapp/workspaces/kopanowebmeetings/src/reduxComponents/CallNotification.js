import {connect} from 'react-redux';

import CallNotification from '@kopanowebmeetings/shared-components/src/CallNotification';
import * as Utils from 'utils/utils.js';

const mapStateToProps = (state, ownProps) => {
	const {calledBy} = state.callNotification;
	return {
		open: state.callNotification.open,
		userName: calledBy ? Utils.getDisplayName(calledBy) : '',
		imgUrl: calledBy ? Utils.imageURLForUser(calledBy) : '',
		onAccept: () => ownProps.onAccept(calledBy),
		onReject: () => ownProps.onReject(calledBy),
	};
};

export default connect(mapStateToProps)(CallNotification);
