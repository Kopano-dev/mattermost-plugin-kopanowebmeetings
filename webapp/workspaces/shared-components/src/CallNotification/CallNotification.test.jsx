import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Configure Enzym to use the react-16 adapter
Enzyme.configure({adapter: new Adapter()});

import CallNotification from './CallNotification.jsx';

describe('CallNotification component', () => {
	test('snapshot tests', () => {
		let component = renderer.create(
			<CallNotification
				open={true}
				userName='Ted Bundy'
				onAccept={() => {}}
				onReject={() => {}}
			/>
		);

		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();

		// Test a component that is not open (=hidden)
		component = renderer.create(
			<CallNotification
				open={false}
				userName='Ted Bundy'
				onAccept={() => {}}
				onReject={() => {}}
			/>
		);

		tree = component.toJSON();
		expect(tree).toMatchSnapshot();

		// Test a component with a title
		component = renderer.create(
			<CallNotification
				open={true}
				title='Calling someone called Ted'
				userName='Ted Bundy'
				onAccept={() => {}}
				onReject={() => {}}
			/>
		);

		tree = component.toJSON();
		expect(tree).toMatchSnapshot();

		// Test a component with an avatar
		component = renderer.create(
			<CallNotification
				open={true}
				imgUrl='https://www.gravatar.com/avatar/3e1b65a10672e6fdf6f376c15a034e87'
				userName='Ted Bundy'
				onAccept={() => {}}
				onReject={() => {}}
			/>
		);

		tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test('accept and reject button', () => {
		let onAcceptCallCount = 0;
		const onAccept = () => onAcceptCallCount++;
		let onRejectCallCount = 0;
		const onReject = () => onRejectCallCount++;
		const component = mount(
			<CallNotification
				open={true}
				userName='Ted Bundy'
				onAccept={onAccept}
				onReject={onReject}
			/>
		);

		component.find('.kwm-accept').simulate('click');
		expect(onAcceptCallCount).toBe(1);

		component.find('.kwm-reject').simulate('click');
		expect(onRejectCallCount).toBe(1);
	});
});