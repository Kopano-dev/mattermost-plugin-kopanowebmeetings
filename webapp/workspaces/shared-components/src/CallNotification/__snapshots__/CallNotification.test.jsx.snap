// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallNotification component snapshot tests 1`] = `
<div
  className="kwm-call-notification"
>
  <div
    className="kwm-call-notification-header"
  >
    <svg
      className="icon-camera-outline"
      style={
        Object {
          "fill": undefined,
        }
      }
      viewBox="0 0 16 12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15.1,1.45,12,3V1a1,1,0,0,0-1-1H1A1,1,0,0,0,0,1V11a1,1,0,0,0,1,1H11a1,1,0,0,0,1-1V9l3.1,1.55A.57.57,0,0,0,16,10V2A.57.57,0,0,0,15.1,1.45ZM12.9,7.21A2,2,0,0,0,12,7H10.5a.5.5,0,0,0-.5.5V10H2V2h8V4.5a.5.5,0,0,0,.5.5H12a2,2,0,0,0,.9-.21L14,4.24V7.76Z"
      />
    </svg>
    Incoming call
  </div>
  <div
    className="kwm-call-notification-body"
  >
    <div
      className="kwm-profile-image-container"
    >
      <div
        className="k-cn-avatar"
      >
        TB
      </div>
    </div>
    <div
      className="kwm-call-notification-msg"
    >
      Ted Bundy
    </div>
  </div>
  <div
    className="kwm-call-buttons"
  >
    <button
      className="kwm-btn kwm-icon-btn kwm-accept"
      onClick={[Function]}
    >
      <svg
        className="icon-pickup"
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20.69,26.56a17.09,17.09,0,0,0,4.09,3.31l2.6-2.36a1.14,1.14,0,0,1,1.18-.21,13.34,13.34,0,0,0,3.57.83,1.15,1.15,0,0,1,.77.37,1.09,1.09,0,0,1,.29.81L33,33.26a1.11,1.11,0,0,1-.37.79,1.14,1.14,0,0,1-.82.29,19.2,19.2,0,0,1-17.74-19.6,1.11,1.11,0,0,1,.37-.79,1.09,1.09,0,0,1,.81-.29l4,.2a1.09,1.09,0,0,1,.78.37,1.11,1.11,0,0,1,.3.8,12.79,12.79,0,0,0,.47,3.63,1.14,1.14,0,0,1-.33,1.15l-2.6,2.36A17.15,17.15,0,0,0,20.69,26.56Z"
        />
      </svg>
    </button>
    <button
      className="kwm-btn kwm-icon-btn kwm-reject"
      onClick={[Function]}
    >
      <svg
        className="icon-hangup"
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M24.35,20.89a14.53,14.53,0,0,0-4.54.72v3.06a1,1,0,0,1-.56.89,12,12,0,0,0-2.63,1.83,1,1,0,0,1-.69.28,1,1,0,0,1-.7-.29l-2.44-2.44a1,1,0,0,1,0-1.41,16.81,16.81,0,0,1,23.13,0,1,1,0,0,1,.29.7,1,1,0,0,1-.29.7l-2.45,2.45a1,1,0,0,1-1.38,0,11.22,11.22,0,0,0-2.64-1.83,1,1,0,0,1-.55-.89V21.6A14.84,14.84,0,0,0,24.35,20.89Z"
        />
      </svg>
    </button>
  </div>
  <audio
    autoPlay={true}
    loop={true}
    src="test-file-stub"
  />
</div>
`;

exports[`CallNotification component snapshot tests 2`] = `null`;

exports[`CallNotification component snapshot tests 3`] = `
<div
  className="kwm-call-notification"
>
  <div
    className="kwm-call-notification-header"
  >
    <svg
      className="icon-camera-outline"
      style={
        Object {
          "fill": undefined,
        }
      }
      viewBox="0 0 16 12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15.1,1.45,12,3V1a1,1,0,0,0-1-1H1A1,1,0,0,0,0,1V11a1,1,0,0,0,1,1H11a1,1,0,0,0,1-1V9l3.1,1.55A.57.57,0,0,0,16,10V2A.57.57,0,0,0,15.1,1.45ZM12.9,7.21A2,2,0,0,0,12,7H10.5a.5.5,0,0,0-.5.5V10H2V2h8V4.5a.5.5,0,0,0,.5.5H12a2,2,0,0,0,.9-.21L14,4.24V7.76Z"
      />
    </svg>
    Calling someone called Ted
  </div>
  <div
    className="kwm-call-notification-body"
  >
    <div
      className="kwm-profile-image-container"
    >
      <div
        className="k-cn-avatar"
      >
        TB
      </div>
    </div>
    <div
      className="kwm-call-notification-msg"
    >
      Ted Bundy
    </div>
  </div>
  <div
    className="kwm-call-buttons"
  >
    <button
      className="kwm-btn kwm-icon-btn kwm-accept"
      onClick={[Function]}
    >
      <svg
        className="icon-pickup"
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20.69,26.56a17.09,17.09,0,0,0,4.09,3.31l2.6-2.36a1.14,1.14,0,0,1,1.18-.21,13.34,13.34,0,0,0,3.57.83,1.15,1.15,0,0,1,.77.37,1.09,1.09,0,0,1,.29.81L33,33.26a1.11,1.11,0,0,1-.37.79,1.14,1.14,0,0,1-.82.29,19.2,19.2,0,0,1-17.74-19.6,1.11,1.11,0,0,1,.37-.79,1.09,1.09,0,0,1,.81-.29l4,.2a1.09,1.09,0,0,1,.78.37,1.11,1.11,0,0,1,.3.8,12.79,12.79,0,0,0,.47,3.63,1.14,1.14,0,0,1-.33,1.15l-2.6,2.36A17.15,17.15,0,0,0,20.69,26.56Z"
        />
      </svg>
    </button>
    <button
      className="kwm-btn kwm-icon-btn kwm-reject"
      onClick={[Function]}
    >
      <svg
        className="icon-hangup"
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M24.35,20.89a14.53,14.53,0,0,0-4.54.72v3.06a1,1,0,0,1-.56.89,12,12,0,0,0-2.63,1.83,1,1,0,0,1-.69.28,1,1,0,0,1-.7-.29l-2.44-2.44a1,1,0,0,1,0-1.41,16.81,16.81,0,0,1,23.13,0,1,1,0,0,1,.29.7,1,1,0,0,1-.29.7l-2.45,2.45a1,1,0,0,1-1.38,0,11.22,11.22,0,0,0-2.64-1.83,1,1,0,0,1-.55-.89V21.6A14.84,14.84,0,0,0,24.35,20.89Z"
        />
      </svg>
    </button>
  </div>
  <audio
    autoPlay={true}
    loop={true}
    src="test-file-stub"
  />
</div>
`;

exports[`CallNotification component snapshot tests 4`] = `
<div
  className="kwm-call-notification"
>
  <div
    className="kwm-call-notification-header"
  >
    <svg
      className="icon-camera-outline"
      style={
        Object {
          "fill": undefined,
        }
      }
      viewBox="0 0 16 12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15.1,1.45,12,3V1a1,1,0,0,0-1-1H1A1,1,0,0,0,0,1V11a1,1,0,0,0,1,1H11a1,1,0,0,0,1-1V9l3.1,1.55A.57.57,0,0,0,16,10V2A.57.57,0,0,0,15.1,1.45ZM12.9,7.21A2,2,0,0,0,12,7H10.5a.5.5,0,0,0-.5.5V10H2V2h8V4.5a.5.5,0,0,0,.5.5H12a2,2,0,0,0,.9-.21L14,4.24V7.76Z"
      />
    </svg>
    Incoming call
  </div>
  <div
    className="kwm-call-notification-body"
  >
    <div
      className="kwm-profile-image-container"
    >
      <img
        src="https://www.gravatar.com/avatar/3e1b65a10672e6fdf6f376c15a034e87"
      />
    </div>
    <div
      className="kwm-call-notification-msg"
    >
      Ted Bundy
    </div>
  </div>
  <div
    className="kwm-call-buttons"
  >
    <button
      className="kwm-btn kwm-icon-btn kwm-accept"
      onClick={[Function]}
    >
      <svg
        className="icon-pickup"
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20.69,26.56a17.09,17.09,0,0,0,4.09,3.31l2.6-2.36a1.14,1.14,0,0,1,1.18-.21,13.34,13.34,0,0,0,3.57.83,1.15,1.15,0,0,1,.77.37,1.09,1.09,0,0,1,.29.81L33,33.26a1.11,1.11,0,0,1-.37.79,1.14,1.14,0,0,1-.82.29,19.2,19.2,0,0,1-17.74-19.6,1.11,1.11,0,0,1,.37-.79,1.09,1.09,0,0,1,.81-.29l4,.2a1.09,1.09,0,0,1,.78.37,1.11,1.11,0,0,1,.3.8,12.79,12.79,0,0,0,.47,3.63,1.14,1.14,0,0,1-.33,1.15l-2.6,2.36A17.15,17.15,0,0,0,20.69,26.56Z"
        />
      </svg>
    </button>
    <button
      className="kwm-btn kwm-icon-btn kwm-reject"
      onClick={[Function]}
    >
      <svg
        className="icon-hangup"
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M24.35,20.89a14.53,14.53,0,0,0-4.54.72v3.06a1,1,0,0,1-.56.89,12,12,0,0,0-2.63,1.83,1,1,0,0,1-.69.28,1,1,0,0,1-.7-.29l-2.44-2.44a1,1,0,0,1,0-1.41,16.81,16.81,0,0,1,23.13,0,1,1,0,0,1,.29.7,1,1,0,0,1-.29.7l-2.45,2.45a1,1,0,0,1-1.38,0,11.22,11.22,0,0,0-2.64-1.83,1,1,0,0,1-.55-.89V21.6A14.84,14.84,0,0,0,24.35,20.89Z"
        />
      </svg>
    </button>
  </div>
  <audio
    autoPlay={true}
    loop={true}
    src="test-file-stub"
  />
</div>
`;
